crates/red_knot_module_resolver/src/resolver.rs

```diff
@@ -143,67 +155,300 @@ impl RawModuleResolutionSettings {
+        // https://docs.python.org/3/library/site.html#module-site
+        let mut all_pth_files: Vec<PthFile> = pth_file_iterator.collect();
+        all_pth_files.sort_by(|a, b| a.path.cmp(&b.path));
+
+        for pth_file in &all_pth_files {
+            dynamic_paths.extend(
+                pth_file
+                    .editable_installations()
+                    .filter_map(|editable_path| {
+                        let possible_search_path = Arc::new(editable_path);
+                        (!static_search_paths.contains(&possible_search_path))
+                            .then_some(possible_search_path)
+                    }),
+            );
         }
     }
+    dynamic_paths
 }
 
-/// A resolved module resolution order as per the [typing spec]
+/// Iterate over the available module-resolution search paths,
+/// following the invariants maintained by [`sys.path` at runtime]:
+/// "No item is added to `sys.path` more than once."
+/// Dynamic search paths (required for editable installs into `site-packages`)
+/// are only calculated lazily.
 ///
-/// [typing spec]: https://typing.readthedocs.io/en/latest/spec/distributing.html#import-resolution-ordering
-#[derive(Clone, Debug, Default, Eq, PartialEq)]
-pub(crate) struct OrderedSearchPaths(Vec<Arc<ModuleResolutionPathBuf>>);
+/// [`sys.path` at runtime]: https://docs.python.org/3/library/site.html#module-site
+struct SearchPathIterator<'db> {
+    db: &'db dyn Db,
+    static_paths: indexmap::set::Iter<'db, SearchPathRoot>,
+    dynamic_paths: Option<indexmap::set::Iter<'db, SearchPathRoot>>,
+}
+
+impl<'db> Iterator for SearchPathIterator<'db> {
+    type Item = &'db SearchPathRoot;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let SearchPathIterator {
+            db,
+            static_paths,
+            dynamic_paths,
+        } = self;
 
-impl Deref for OrderedSearchPaths {
-    type Target = [Arc<ModuleResolutionPathBuf>];
+        static_paths.next().or_else(|| {
+            dynamic_paths
+                .get_or_insert_with(|| dynamic_module_resolution_paths(*db).into_iter())
+                .next()
+        })
+    }
+}
+
+impl<'db> FusedIterator for SearchPathIterator<'db> {}
 
-    fn deref(&self) -> &Self::Target {
-        &self.0
+/// Represents a single `.pth` file in a `site-packages` directory.
+/// One or more lines in a `.pth` file may be a (relative or absolute)
+/// path that represents an editable installation of a package.
+struct PthFile<'db> {
+    system: &'db dyn System,
+    path: SystemPathBuf,
+    contents: SourceText,
+    site_packages: &'db SystemPath,
+}
+
+impl<'db> PthFile<'db> {
+    /// Yield paths in this `.pth` file that appear to represent editable installations,
+    /// and should therefore be added as module-resolution search paths.
+    fn editable_installations<'a>(&'a self) -> impl Iterator<Item = ModuleResolutionPathBuf> + 'db
+    where
+        'a: 'db,
+    {
```

Comment:
Nit: I think you can just use the `db` lifetime
```suggestion
    fn editable_installations<'a>(&'db self) -> impl Iterator<Item = ModuleResolutionPathBuf> + 'db
    {
```